### PR TITLE
LPD-26191 Adde virtual Donwload Section

### PR DESCRIPTION
**Pull request by welitonpe**

### CODE REVIEW

**Comment by welitonpe** [comment]
ci:test:sf

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/components/PurchasedAppsTable.tsx [hunk 1]
```diff
@@ -237,23 +238,37 @@ const AppsTable: React.FC<AppsTableProps> = ({items}) => {
...
 													onClick={() => {
-														if (!virtualURL) {
+														if (
+															properties.featureFlags?.includes(
+																'LPD-21582'
+															) &&
+															!isFreeApp
+														) {
+															Liferay.Util.navigate(
+																`${Liferay.ThemeDisplay.getLayoutURL()}#order/${id}/download`
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/components/PurchasedAppsTable.tsx:248]: Use a relative path for this

```suggestion
																`order/${id}/download`
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/AppOutlet.tsx [hunk 2]
```diff
@@ -96,6 +117,12 @@ const AppOutlet = () => {
 				/>
 
 				<AppNavbar
+					showDownloadTab={
+						data?.placedOrder.workflowStatusInfo.code === 0 &&
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/AppOutlet.tsx:121]: Use enum

```suggestion
						data?.placedOrder.workflowStatusInfo.code === 0 &&
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/Download.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+import {useMarketplaceContext} from '../../../../../../context/MarketplaceContext';
+import useGetProductByOrderId from '../../../../../../hooks/useGetProductByOrderId';
+import HeadlessCommerceDeliveryCatalogImpl from '../../../../../../services/rest/HeadlessCommerceDeliveryCatalog';
+import DownloadTable from './DownloadTable';
+
+type OutletContext = ReturnType<typeof useGetProductByOrderId>;
+
+const Download = () => {
+	const marketplaceContetx = useMarketplaceContext();
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/Download.tsx:20]:
```suggestion
	const marketplaceContext = useMarketplaceContext();
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/DownloadTable.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import ClayButton from '@clayui/button';
+
+import EmptyState from '../../../../../../components/EmptyState';
+import Loading from '../../../../../../components/Loading';
+import Table from '../../../../../../components/Table/Table';
+import i18n from '../../../../../../i18n';
+
+type DownloadTableProps = {
+	loading: boolean;
+	virualItems: VirtualItem[];
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/DownloadTable.tsx:15]:
```suggestion
	virtualItems: VirtualItem[];
```

**Review by kevenleone — CHANGES_REQUESTED** [review]

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/components/PurchasedAppsTable.tsx [hunk 5]
```diff
@@ -237,23 +238,37 @@ const AppsTable: React.FC<AppsTableProps> = ({items}) => {
...
+																'DOWNLOAD_APP',
 																metadata
 															);
 
-															return alert(
-																'Download file not found'
+															window.open(
+																virtualURL
 															);
 														}
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/components/PurchasedAppsTable.tsx:271]:
```suggestion
														 if (properties.featureFlags?.includes('LPD-21582') && !isFreeApp) {
    const downloadURL = `${Liferay.ThemeDisplay.getLayoutURL()}#order/${id}/download`;
    Liferay.Util.navigate(downloadURL);
    return;
  }

  if (!virtualURL) {
    Analytics.track('VIRTUAL_URL_NOT_FOUND', metadata);
    alert('Download file not found');
    return;
  }

  Analytics.track('DOWNLOAD_APP', metadata);
  window.open(virtualURL);
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/AppOutlet.tsx [hunk 6]
```diff
@@ -70,6 +90,7 @@ const AppOutlet = () => {
 	const {isFreeApp} = getProductPriceModel(product);
 	const navigate = useNavigate();
 	const placedOrderItems = data?.placedOrder.placedOrderItems ?? [];
+
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/AppOutlet.tsx:93]: is this blank line mandatory?

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/AppOutlet.tsx [hunk 7]
```diff
@@ -96,6 +117,12 @@ const AppOutlet = () => {
 				/>
 
 				<AppNavbar
+					showDownloadTab={
+						data?.placedOrder.workflowStatusInfo.code === 0 &&
+						data?.placedOrder.placedOrderItems.some(
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/AppOutlet.tsx:122]:
```suggestion
						placedOrderItems.some(
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/Download.tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+	);
+
+	const virualItems = useMemo(() => {
+		const hasSkuVersion = skus[0]?.customFields.find(
+			(customField: CustomField) =>
+				customField.name === 'Version' && customField.customValue.data
+		);
+
+		const virtualItemsWithVersion = virtualProducts[0].virtualItems?.map(
+			(virtualItem: any) => ({
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/Download.tsx:57]: Can we remove this 'any'?

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/Download.tsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+				...virtualItem,
+				productVersion: hasVersionSpecification
+					? hasVersionSpecification.value
+					: hasSkuVersion?.customValue.data,
+			})
+		);
+
+		return virtualItemsWithVersion.filter((item: any) =>
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/CustomerDashboard/pages/Apps/App/Download/Download.tsx:65]: same

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/NextSteps/NextSteps.tsx [hunk 10]
```diff
@@ -224,13 +233,31 @@ export function NextSteps({
...
+						}
+
+						if (
+							cart.orderTypeExternalReferenceCode === 'CLOUDAPP'
+						) {
+							if (onClickContinue) {
+								window.location.href =
+									'https://console.liferay.cloud/projects';
+							}
 						}
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/NextSteps/NextSteps.tsx:255]:
```suggestion
						
if (
  cart.orderTypeExternalReferenceCode === 'CLOUDAPP' &&
  onClickContinue
) {
  window.location.href = 'https://console.liferay.cloud/projects';
}
```

**Review by joanagrego — CHANGES_REQUESTED** [review]

**Comment by welitonpe** [comment]
ci:test:sf

**Review by kevenleone — APPROVED** [review]
LGTM 🔥 

Good job @welitonpe

**Comment by kevenleone** [comment]
ci:forward

**Comment by kevenleone** [comment]
ci:stop

**Comment by kevenleone** [comment]
Just starting reviewing :)

**Pull request closed without merging**